dcicutils/obfuscation_utils.py

```diff
@@ -96,25 +154,30 @@ def obfuscate_dict(dictionary: dict, inplace: bool = False, show: bool = False,
     check_true(not obfuscated or is_obfuscated(obfuscated),
                message=f"If obfuscated= is supplied, it must be {OBFUSCATED_VALUE_DESCRIPTION}.")
 
-    def has_values_to_obfuscate(dictionary: dict) -> bool:
-        for key, value in dictionary.items():
-            if isinstance(value, dict):
-                if has_values_to_obfuscate(value):
-                    return True
-            elif isinstance(value, str) and should_obfuscate(key) and not is_obfuscated(value):
-                return True
-        return False
+    if show:
+        return item
+
+    orig_item = item
+
+    changed = StorageCell(False)
```

Comment:
Oh, weird, I've only just not learned (sort of) why this `changed` variable needs to be a class/object; not intuitive to me; so when `process_recursively` is called recursively, somehow if it is a value (non-reference) type (like just bool) then it would be a new instance of it; could maybe use some good Kent-style pedagogy on this. (ChatGPT is trying to explain it, and experimenting with it, but still not 100% there in my understanding).

Reply:
The problem is that in any function, if you assign a variable, that variable becomes local to the function. You can read the value of an outer variable, but you cannot write one. (In fact, in some version of Python, I'm not sure which, there is a way of declaring (similar to `global`) that you mean to share the variables between the contexts so the assignment should happen farther out, but since `dcicutils` spans several versions, I didn't do that. And in any case this StorageCell thing can be used for many such things.

No, it is not creating a new instance on recursive calls. The only instance creation is outside the function, so it is a shared single location that all locations see and asks the question, "was there any changed value at all?"  I could, alternatively, just always copy. But this satisfies the previous unit tests, most of them.  (There is also a way to efficiently avoid having to do this which I can show you, but it was late and I was tired so I did not use that technique because I've never done it in Python and I'd have to puzzle it out a little more than I would in Lisp. I did change the unit test so that if something "looked obfuscated", it still ends up thinking it's copied it when perhaps it hasn't.)

Reply:
Thanks for that. Looks like could also use the `nonlocal` keyword feature.